vocabularies/Common.json

```diff
@@ -1288,6 +1288,132 @@
+      "@Common.Experimental": true,
+      "@Core.Description": "Qualified name of a function that computes AI-based recommendations for entities of the annotated type (see [template](#Template_RecommendationsFunction) and [template for asynchronous OData requests](#Template_AsyncRecommendationsFunction))"
+    },
+    "Template_RecommendationsFunction": [
+      {
+        "$Kind": "Function",
+        "$IsBound": true,
+        "@Common.Experimental": true,
+        "@Core.Description": "Template for functions that compute AI-based recommendations for the bound entity and its related entities",
+        "@Core.LongDescription": "If the function is invoked repeatedly in a stateful session, it SHOULD NOT\nrecompute recommendations whose predictors have not changed since the previous invocation.\n\nThe template function itself cannot be invoked.",
+        "$Parameter": [
+          { "$Name": "Entity", "$Type": "Edm.EntityType" },
+          {
+            "$Name": "Targets",
+            "$Collection": true,
+            "@Core.IsURL": true,
+            "@Core.Description": "URLs (relative to the request URL) that address entities, properties or action/function parameters for which recommendations shall be retrieved",
+            "@Core.LongDescription": "The request `SalesOrder('A')/ns.RecommendationsFunction?Targets=[\"Items(10)\"]`\n            retrieves recommendations for the entity `SalesOrder('A')/Items(10)`."
+          }
+        ],
+        "$ReturnType": {
+          "@Core.IsURL": true,
+          "@Core.Description": "URL (relative to the service URL) for retrieving the recommendations (which are computed asynchronously)",
+          "@Core.Example": {
+            "@odata.type": "https://oasis-tcs.github.io/odata-vocabularies/vocabularies/Org.OData.Core.V1.xml#Core.PrimitiveExampleValue",
+            "Description": "Relative URL for invoking a function import of the [retrieval function](#Template_RecommendationsRetrievalFunction)",
+            "Value": "RetrieveRecommendations?ID=af43c2"
+          }
+        }
+      }
+    ],
+    "Template_RecommendationsRetrievalFunction": [
+      {
+        "$Kind": "Function",
+        "@Common.Experimental": true,
+        "@Core.Description": "Template for functions that retrieve AI-based recommendations for the bound entity and its related entities",
+        "@Core.LongDescription": "The template function itself cannot be invoked.",
+        "$Parameter": [
+          {
+            "$Name": "ID",
+            "@Core.Description": "ID that occurs in the URL returned by a prior invocation of the [recommendations function](#Template_RecommendationsFunction)"
+          }
+        ],
+        "$ReturnType": {
+          "$Collection": true,
+          "$Type": "Common.PropertyRecommendationType",
+          "@Core.Description": "A collection of recommendations per targeted property"
+        }
+      }
+    ],
+    "Template_AsyncRecommendationsFunction": [
```

Comment:
Even thought it is clear to me, we could ignore till the framework implement the support for recommendations correctly, isn't it?

Reply:
Yes, this template becomes relevant only when Gateway supports this type of asynchronous requests.

Reply:
There is only the `RecommendationsFunction` annotation and it will either list a function following the `` `Template_RecommendationsFunction` or the `Template_AsyncRecommendationsFunction`. FE will need to decide which one is called by looking at the return type? Or will there be another way of distinguishing the two?

Reply:
Aren't the name for the template incorrect ?

The Template_RecommendationsFunction returns a url where the data can finally be retrieved while this one
Template_AsyncRecommendationsFunction returns the data 

For me the `async` component of all this relates to the asynchronous requests part....
Or are you saying that at no point in time do you consider that the recommendation service can directly return you the recommendation data ?
In case the data is cached this should be a pretty fast call...

Reply:
If the return type is `Edm.String`, FE calls the function and polls the result at the URL returned by the function call.
Otherwise, FE calls the function with `Prefer: respond-async` and polls the result at the `Location` header returned by that request.
Is that acceptable?

Reply:
That i understand yes, but why not also support a simpler case where you already get the `Collection(Common.PropertyRecommendationType)`
In repeat case where the response would already be the same and cached (as indicated in your recommendation), we would avoid an extra call

Reply:
@nlunets Using asynchronous requests would also allow that use case, at least in theory. The server may or may not apply the preference, i.e. it could decide whether the request can be responded to quickly and synchronously or not. However, my understanding is that the way GW works, such a distinction will not be possible.

To add another function that always returns the results synchronously would be possible. But how would FE know that it can call this function? The solution sketched in the spec for asynchronous requests, i.e. that the server decides whether to respond asynchronously, seem superior to me.

Reply:
In the http world, asynchronous request with 202 are mostly designed for _long_ running jobs, which can range from tens of second to tens of minutes.
If we already consider that recommendation are going to take that long to be returned, we might as well not work on them :)

I'm not against considering that exceptionally, a service might decide to return them as an async call, but i would consider that most case should be implemented as a direct response.

For FE that's rather easy i'd say:

- If the function has a returnType of recommendation Common.PropertyRecommendationType -> it's a direct response one
- If the function has a returnType of type String -> by convention we can poll
- If the function is annotated with something to indicate it's an asynchronous process on the backend -> Async Request handling

Reply:
The simpler case will be supported once we switch to `Prefer: respond-async`. Then the server can decide whether to respond synchronously or asynchronously. But this is not yet supported by Gateway.

Until then, an additional second request is better than a slower first request.

Reply:
The server can decide. However, the recommendations call has to be requested only after the application of the PATCH. There are two ways how to achieve that. Either it is a sequential request after the $batch with the PATCH or it is part of the same $batch. In the latter case, there is typically also a side effects GET. So if fetching the recommendations takes several seconds, the user would see this as a slower application of side effects.
And with sticky there seems to be the restriction that only one request is processed at a time. And the recommendations call needs to use the sticky session. 
So overall, becoming asynchronous, even if it is for just a few seconds, seems to make sense to not slow down the normal processing with PATCHes and side effects.